feat(CategoryTheory/Adjunction): more results on adjoint triples.

peabrainiac opened the pull request:
Lemmas about adjoint triples `F ⊣ G ⊣ H` where either `G` is fully faithful or `F` and `H` are.

From [lean-orbifolds](https://github.com/peabrainiac/lean-orbifolds/blob/387fe778f3cc1ed3c5e60f390842333ecc9eda67/Orbifolds/ForMathlib/Triple.lean).

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

peabrainiac committed:
copy over code from the orbifolds repo

peabrainiac committed:
move lemmas on epimorphisms and monomorphisms

peabrainiac committed:
shuffle around lemmas, update module docstring

peabrainiac commented:
I think the large import is nothing to worry about, since this is a leaf file and it is not clear yet if any future importers would benefit from splitting it up; however, since it is only 8 lemmas that need the additional import it could also make sense to move them into some new file like `Adjunction.EpiMono`. I plan on adding two more lemmas like that in another PR later, on quadruples instead of triples.

Particular attention should probably go to the lemmas added to `CategoryTheory.EpiMono`, since they are general API rather than specific to adjoint triples; let me know if I should split them off into a separate PR.

peabrainiac committed:
try tagging a few lemmas with `simp`

peabrainiac commented:
!bench

joelriou commented on Mathlib/CategoryTheory/EpiMono.lean:
```diff
@@ -242,4 +242,68 @@ instance {X Y : C} (f : X ⟶ Y) [hf : IsSplitEpi f] (F : C ⥤ D) : IsSplitEpi
+
+/-- When `f` is an isomorphism, `f ≫ g` is epic iff `g` is. -/
+@[simp]
+lemma epi_isIso_comp_iff {X Y Z : C} (f : X ⟶ Y) [IsIso f] (g : Y ⟶ Z) :
```
There is a more general lemma:
```lean
lemma epi_comp_iff {X Y Z : C} (f : X ⟶ Y) [Epi f] (g : Y ⟶ Z) :
    Epi (f ≫ g) ↔ Epi g :=
  ⟨fun _ ↦ epi_of_epi f g, fun _ ↦ inferInstance⟩
```

joelriou commented on Mathlib/CategoryTheory/Adjunction/Triple.lean:
```diff
@@ -52,4 +82,226 @@ noncomputable def fullyFaithfulEquiv : F.FullyFaithful ≃ H.FullyFaithful where
+/-- For an adjoint triple `F ⊣ G ⊣ H`, the components of the natural transformation
+`H ⋙ G ⟶ F ⋙ G` obtained from the units and counits of the adjunctions are
+under the second adjunction adjunct to the image of the unit of the first adjunction under `H`. -/
+lemma homEquiv_counit_unit_app_eq_H_map_unit {X : C} :
```
We should not put letter like `F`, `G`, `H` in names. For convenience, I would suggest introducing a structure as follows, which also makes it easier to get reasonable names which follow the naming conventions:
```lean
variable (F : C ⥤ D) (G : D ⥤ C) (H : C ⥤ D)

structure Triple where
  adj₁ : F ⊣ G
  adj₂ : G ⊣ H

namespace Triple

variable {F G H} (t : Triple F G H) [G.Full] [G.Faithful]

noncomputable def natTrans : H ⟶ F :=
  ((Functor.FullyFaithful.ofFullyFaithful G).whiskeringRight _).preimage
    (t.adj₂.counit ≫ t.adj₁.unit)

@[simp, reassoc]
lemma whiskerRight_natTrans : whiskerRight t.natTrans G = t.adj₂.counit ≫ t.adj₁.unit :=
  ((Functor.FullyFaithful.ofFullyFaithful G).whiskeringRight _).map_preimage _

@[simp, reassoc]
lemma map_natTrans_app (X : C) :
    G.map (t.natTrans.app X) =
      t.adj₂.counit.app X ≫ t.adj₁.unit.app X :=
  congr_app t.whiskerRight_natTrans X

@[reassoc (attr := simp)]
lemma adj₁_counit_app_natTrans_app (X : C) :
    t.adj₁.counit.app (H.obj X) ≫ t.natTrans.app X = F.map (t.adj₂.counit.app X) :=
  G.map_injective (by simp [← cancel_epi (t.adj₁.unit.app _)])

@[reassoc (attr := simp)]
lemma natTrans_app_adj₂_unit_app (X : C) :
    t.natTrans.app X ≫ t.adj₂.unit.app (F.obj X) = H.map (t.adj₁.unit.app X) :=
  G.map_injective (by simp [← cancel_mono (t.adj₂.counit.app _)])

lemma epi_natTrans_app_iff_epi_map_adj₂_counit_app (X : C) :
    Epi (t.natTrans.app X) ↔ Epi (F.map (t.adj₂.counit.app X)) := by
  rw [← epi_comp_iff (t.adj₁.counit.app (H.obj X)), adj₁_counit_app_natTrans_app]

lemma epi_natTrans_app_iff (X : C) :
    Epi (t.natTrans.app X) ↔ Epi (H.map (t.adj₁.unit.app X)) := by
  rw [← epi_comp_isIso_iff _ (t.adj₂.unit.app (F.obj X)), natTrans_app_adj₂_unit_app]

end Triple
```
I am not completely sure we need to state the lemmas about natural transformations being `Mono/Epi`. I would suggest keeping in this PR only the `_app` versions, and if we absolutely need to use `Epi/Mono` natural transformations in applications, we may use `NatTrans.epi_iff_epi_app` there only.

peabrainiac replied:
What would you suggest calling the natural transformation `H ⟶ F` if the one going the other direction is `Triple.natTrans`? Maybe `Triple.leftToRight` and `Triple.rightToLeft` could work?

joelriou commented on Mathlib/CategoryTheory/EpiMono.lean:
```diff
@@ -242,4 +242,68 @@ instance {X Y : C} (f : X ⟶ Y) [hf : IsSplitEpi f] (F : C ⥤ D) : IsSplitEpi
+
+/-- For any natural transformation `f : F ⟶ G`, if `X` and `Y` are isomorphic, the component of
+`f` at `X` is epic iff the component of `f` at `Y` is. -/
+lemma NatTrans.epi_app_congr_iso {C D : Type*} [Category C] [Category D]
```
As I am not completely use we absolutely need these lemmas (in my code above about `HToF`, I did not need it), I would suggest keeping these lemmas out of this PR.

peabrainiac committed:
introduce `Triple` structure

peabrainiac committed:
rename `HToF` -> `rightToLeft`, `FToH` -> `leftToRight`

peabrainiac committed:
remove lemmas about unapplied natural transfromations being epic/monic

peabrainiac committed:
apply more suggestions

peabrainiac committed:
fix

peabrainiac committed:
fix comments

peabrainiac committed:
remove accidentally included line

peabrainiac commented:
I've renamed the natural transformations and introduced the `Triple` structure now. I'm a bit unsure about the latter change; it definitely solves the naming problem, but also adds complexity in that when you want to apply the lemmas to adjunctions `adj₁`, `adj₂`, you now need to construct the triple `⟨t.adj₁, t.adj₂⟩` first, and then the lemmas are about `⟨t.adj₁, t.adj₂⟩.adj₁` and `⟨t.adj₁, t.adj₂⟩.adj₂` instead of `adj₁` and `t.adj₂`. Of course that's equal, but not reducibly so; I'm not sure if that could cause issues.

Regarding the lemmas about natural transformations as a whole being epic/monic, you're probably right that just proving that when needed is enough; I've removed them now. That also had the nice side effect of removing the extra imports, so the question of whether those lemmas should go into a different file is resolved now too.

That is all I can do for now; I'll have a think about the alternative structure you posted later.

peabrainiac committed:
refactor

peabrainiac committed:
update module docstring

peabrainiac committed:
remove simps projection

peabrainiac committed:
remove redundant simp tag

peabrainiac commented:
I have implemented your alternative definition `rightToLeft` now; it was indeed quite helpful for cleaning up the proofs. I also tried to clean up `leftToRight` in a similar way, but as far as I can see this approach doesn't really carry over to that case, so there is somewhat of an asymmetry in the API now. Nonetheless, I think this is as good as I can make it for now.

peabrainiac commented on Mathlib/CategoryTheory/Adjunction/Triple.lean:
```diff
@@ -38,18 +76,186 @@ noncomputable def fullyFaithfulEquiv : F.FullyFaithful ≃ H.FullyFaithful where
+/-- The natural transformation `F ⟶ H` that exists for every adjoint triple `F ⊣ G ⊣ H` where `F`
+and `H` are fully faithful, given here as the whiskered unit `F ⟶ F ⋙ G ⋙ H` of the second
+adjunction followed by the inverse of the whiskered unit `F ⋙ G ⋙ H ⟶ H` of the first. -/
+noncomputable def leftToRight : F ⟶ H :=
+  F.rightUnitor.inv ≫ whiskerLeft F t.adj₂.unit ≫ (Functor.associator _ _ _).inv ≫
+  inv (whiskerRight t.adj₁.unit H) ≫ H.leftUnitor.hom
```
Are you sure we want this, considering that `F` is fully faithful iff `H` is? Both being fully faithful is always just one application of `Triple.fullyFaithfulEquiv` away. In my mind, we should have both instances registered anyway in all situations where these lemmas would be applied.
